File: web/src/sim/analyzer.ts
// Port of py/diagkit_rca/analyzer.py: rank services by likelihood of being the
// incident root cause. A service scores high when three independent signals
// line up: (1) a dense cluster of recurring error signatures, (2) an error-rate
// and/or p95 latency spike over the window, and (3) other services' errors
// propagating through it. The score is an explainable weighted sum of those
// three normalized signals.

import type { Bundle } from "./types";

// Signal weights, identical to the analyzer. Signature density and the metric
// spike are the strongest evidence; propagation breaks ties toward the true
// upstream culprit.
export const W_SIGNATURE = 0.4;
export const W_SPIKE = 0.35;
export const W_PROPAGATION = 0.25;

export interface RootCause {
  service: string;
  score: number;
  signatureCount: number;
  signatureErrors: number;
  latencySpikeX: number;
  errorRatePeak: number;
  propagationPct: number;
  reasons: string[];
}

function round(v: number, digits: number): number {
  const f = 10 ** digits;
  return Math.round((v + Number.EPSILON) * f) / f;
}

function signatureSignal(b: Bundle): [Record<string, number>, Record<string, number>] {
  const clusters: Record<string, number> = {};
  const errors: Record<string, number> = {};
  for (const sig of b.signatures) {
    for (const svc of sig.services) {
      clusters[svc] = (clusters[svc] ?? 0) + 1;
      errors[svc] = (errors[svc] ?? 0) + sig.count;
    }
  }
  return [clusters, errors];
}

function metricSpikes(b: Bundle): [Record<string, number>, Record<string, number>] {
  const latencyX: Record<string, number> = {};
  const errPeak: Record<string, number> = {};
  for (const m of b.metrics) {
    const lat = m.p95_latency_ms.map((p) => p.value);
    const values = lat.length ? lat : [0.0];
    const positive = values.filter((v) => v > 0);
    const baseline = positive.length ? Math.min(...positive) : 1.0;
    latencyX[m.service] = baseline > 0 ? Math.max(...values) / baseline : 1.0;
    errPeak[m.service] = m.error_rate.length ? Math.max(...m.error_rate.map((p) => p.value)) : 0.0;
  }
  return [latencyX, errPeak];
}

function propagation(b: Bundle): Record<string, number> {
  const entryCandidates = [...new Set(b.traces.filter((s) => !s.called_by).map((s) => s.service))].sort();
  const entry = entryCandidates.length ? entryCandidates[0] : b.services.length ? b.services[0] : "";

  const entryErrorTraces = new Set<string>();
  const svcErrorTraces: Record<string, Set<string>> = {};
  for (const span of b.traces) {
    if (span.error) {
      (svcErrorTraces[span.service] ??= new Set()).add(span.trace_id);
      if (span.service === entry) entryErrorTraces.add(span.trace_id);
    }
  }

  const total = entryErrorTraces.size;
  const pct: Record<string, number> = {};
  for (const svc of b.services) {
    if (total === 0) {
      pct[svc] = 0.0;
    } else {
      const set = svcErrorTraces[svc] ?? new Set();
      let overlap = 0;
      for (const t of set) if (entryErrorTraces.has(t)) overlap++;
      pct[svc] = overlap / total;
    }
  }
  return pct;
}

function normalizeSignal(values: Record<string, number>): Record<string, number> {
  const vals = Object.values(values);
  const hi = vals.length ? Math.max(...vals) : 0.0;
  const out: Record<string, number> = {};
  if (hi <= 0) {
    for (const k of Object.keys(values)) out[k] = 0.0;
    return out;
  }
  for (const [k, v] of Object.entries(values)) out[k] = v / hi;
  return out;
}

function reasons(
  svc: string,
  clusters: Record<string, number>,
  sigErrors: Record<string, number>,
  latencyX: Record<string, number>,
  errPeak: Record<string, number>,
  prop: Record<string, number>
): string[] {
  const out: string[] = [];
  if ((clusters[svc] ?? 0) > 0) {
    out.push(`${clusters[svc]} error signature(s) covering ${sigErrors[svc] ?? 0} log lines`);
  }
  const lx = latencyX[svc] ?? 1.0;
  if (lx >= 1.5) out.push(`p95 latency spike ${lx.toFixed(1)}x baseline`);
  const ep = errPeak[svc] ?? 0.0;
  if (ep >= 0.1) out.push(`error rate peaked at ${Math.round(ep * 100)}%`);
  const p = prop[svc] ?? 0.0;
  if (p >= 0.1) out.push(`${Math.round(p * 100)}% of entry errors trace through it`);
  if (out.length === 0) out.push("no significant failure signal");
  return out;
}

// analyze ranks services by likelihood of being the incident root cause.
export function analyze(b: Bundle): RootCause[] {
  const [clusters, sigErrors] = signatureSignal(b);
  const [latencyX, errPeak] = metricSpikes(b);
  const prop = propagation(b);

  const spikeRaw: Record<string, number> = {};
  for (const svc of b.services) {
    spikeRaw[svc] = (latencyX[svc] ?? 1.0) - 1.0 + (errPeak[svc] ?? 0.0) * 5.0;
  }

  const nSigInput: Record<string, number> = {};
  const nSpikeInput: Record<string, number> = {};
  const nPropInput: Record<string, number> = {};
  for (const svc of b.services) {
    nSigInput[svc] = sigErrors[svc] ?? 0;
    nSpikeInput[svc] = Math.max(0.0, spikeRaw[svc]);
    nPropInput[svc] = prop[svc] ?? 0.0;
  }
  const nSig = normalizeSignal(nSigInput);
  const nSpike = normalizeSignal(nSpikeInput);
  const nProp = normalizeSignal(nPropInput);

  const results: RootCause[] = b.services.map((svc) => {
    const score = W_SIGNATURE * nSig[svc] + W_SPIKE * nSpike[svc] + W_PROPAGATION * nProp[svc];
    return {
      service: svc,
      score: round(score, 4),
      signatureCount: clusters[svc] ?? 0,
      signatureErrors: sigErrors[svc] ?? 0,
      latencySpikeX: round(latencyX[svc] ?? 1.0, 2),
      errorRatePeak: round(errPeak[svc] ?? 0.0, 3),
      propagationPct: round((prop[svc] ?? 0.0) * 100, 1),
      reasons: reasons(svc, clusters, sigErrors, latencyX, errPeak, prop),
    };
  });

  results.sort((a, b2) => {
    if (a.score !== b2.score) return b2.score - a.score;
    if (a.signatureErrors !== b2.signatureErrors) return b2.signatureErrors - a.signatureErrors;
    return a.service < b2.service ? 1 : a.service > b2.service ? -1 : 0;
  });
  return results;
}

File: py/diagkit_rca/analyzer.py
"""Root-cause ranking.

A service scores high as a likely root cause when three independent signals
line up:

1. it owns a dense cluster of recurring error signatures,
2. its metrics show an error-rate and/or p95 latency spike over the window, and
3. errors in other services propagate through it (dependency blast radius).

The score is an explainable weighted sum of those three normalized signals, so
every ranking can be justified in plain language.
"""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field

from .bundle import Bundle

# Signal weights. Signature density and the metric spike are the strongest
# evidence; propagation breaks ties toward the true upstream culprit.
W_SIGNATURE = 0.4
W_SPIKE = 0.35
W_PROPAGATION = 0.25


@dataclass
class RootCause:
    service: str
    score: float
    signature_count: int
    signature_errors: int
    latency_spike_x: float
    error_rate_peak: float
    propagation_pct: float
    reasons: list[str] = field(default_factory=list)


def _signature_signal(bundle: Bundle) -> tuple[dict[str, int], dict[str, int]]:
    """Return per-service signature cluster counts and total matched errors."""
    clusters: dict[str, int] = defaultdict(int)
    errors: dict[str, int] = defaultdict(int)
    for sig in bundle.signatures:
        for svc in sig.services:
            clusters[svc] += 1
            errors[svc] += sig.count
    return clusters, errors


def _metric_spikes(bundle: Bundle) -> tuple[dict[str, float], dict[str, float]]:
    """Return per-service latency spike multiplier and peak error rate."""
    latency_x: dict[str, float] = {}
    err_peak: dict[str, float] = {}
    for m in bundle.metrics:
        lat = [p.value for p in m.p95_latency_ms] or [0.0]
        baseline = min(v for v in lat if v > 0) if any(v > 0 for v in lat) else 1.0
        latency_x[m.service] = (max(lat) / baseline) if baseline > 0 else 1.0
        err_peak[m.service] = max((p.value for p in m.error_rate), default=0.0)
    return latency_x, err_peak


def _propagation(bundle: Bundle) -> dict[str, float]:
    """Fraction of the entry service's error traces that pass through each service.

    The entry service is the one that is never called by anyone. A downstream
    service with high coverage of the entry service's failing traces is a strong
    root-cause candidate.
    """
    called_by = {s.called_by for s in bundle.traces}
    entry_candidates = [s for s in bundle.services if s not in called_by]
    if entry_candidates:
        entry = entry_candidates[0]
    else:
        entry = bundle.services[0] if bundle.services else ""

    entry_error_traces: set[str] = set()
    svc_error_traces: dict[str, set[str]] = defaultdict(set)
    for span in bundle.traces:
        if span.error:
            svc_error_traces[span.service].add(span.trace_id)
            if span.service == entry:
                entry_error_traces.add(span.trace_id)

    total = len(entry_error_traces)
    pct: dict[str, float] = {}
    for svc in bundle.services:
        if total == 0:
            pct[svc] = 0.0
        else:
            overlap = len(svc_error_traces[svc] & entry_error_traces)
            pct[svc] = overlap / total
    return pct


def _normalize(values: dict[str, float]) -> dict[str, float]:
    hi = max(values.values(), default=0.0)
    if hi <= 0:
        return {k: 0.0 for k in values}
    return {k: v / hi for k, v in values.items()}


def analyze(bundle: Bundle) -> list[RootCause]:
    """Rank services by likelihood of being the incident root cause."""
    clusters, sig_errors = _signature_signal(bundle)
    latency_x, err_peak = _metric_spikes(bundle)
    prop = _propagation(bundle)

    # Combine latency and error-rate into one spike signal per service.
    spike_raw = {
        svc: (latency_x.get(svc, 1.0) - 1.0) + err_peak.get(svc, 0.0) * 5.0
        for svc in bundle.services
    }

    n_sig = _normalize({svc: float(sig_errors.get(svc, 0)) for svc in bundle.services})
    n_spike = _normalize({svc: max(0.0, v) for svc, v in spike_raw.items()})
    n_prop = _normalize({svc: prop.get(svc, 0.0) for svc in bundle.services})

    results: list[RootCause] = []
    for svc in bundle.services:
        score = (
            W_SIGNATURE * n_sig[svc] + W_SPIKE * n_spike[svc] + W_PROPAGATION * n_prop[svc]
        )
        rc = RootCause(
            service=svc,
            score=round(score, 4),
            signature_count=clusters.get(svc, 0),
            signature_errors=sig_errors.get(svc, 0),
            latency_spike_x=round(latency_x.get(svc, 1.0), 2),
            error_rate_peak=round(err_peak.get(svc, 0.0), 3),
            propagation_pct=round(prop.get(svc, 0.0) * 100, 1),
            reasons=_reasons(svc, clusters, sig_errors, latency_x, err_peak, prop),
        )
        results.append(rc)

    results.sort(key=lambda r: (r.score, r.signature_errors, r.service), reverse=True)
    return results


def _reasons(
    svc: str,
    clusters: dict[str, int],
    sig_errors: dict[str, int],
    latency_x: dict[str, float],
    err_peak: dict[str, float],
    prop: dict[str, float],
) -> list[str]:
    out: list[str] = []
    if clusters.get(svc, 0) > 0:
        out.append(
            f"{clusters[svc]} error signature(s) covering {sig_errors.get(svc, 0)} log lines"
        )
    lx = latency_x.get(svc, 1.0)
    if lx >= 1.5:
        out.append(f"p95 latency spike {lx:.1f}x baseline")
    ep = err_peak.get(svc, 0.0)
    if ep >= 0.1:
        out.append(f"error rate peaked at {ep * 100:.0f}%")
    p = prop.get(svc, 0.0)
    if p >= 0.1:
        out.append(f"{p * 100:.0f}% of entry errors trace through it")
    if not out:
        out.append("no significant failure signal")
    return out
